### ISSUE TITLE
Handle PID changes during reload

### ISSUE THREAD

**Issue by bodymindarts** [habitat-sh/habitat#1838]
This is a follow up issue to https://github.com/habitat-sh/habitat/issues/1801 which is handled for the most part by https://github.com/habitat-sh/habitat/pull/1837.

In some cases the native `reload` support for an application would actually change the `PID` of the process (eg. haproxy). The current implementation of the supervisor however cannot support a change of `PID`.

**Comment by bodymindarts** [comment]
From a comment by @reset in https://github.com/habitat-sh/habitat/issues/1801

>A simple hook to send a reload command or run a binary to communicate to a PID is definitely in the scope of how the supervisor works right now, but if any sort of process draining/swapping needs to happen where the PID will change, we don't have a way to represent that just yet. The supervisor starts a process as a child of itself and attempts to restart it if that PID ever goes away. Simply put, there's no way for a process' PID to change as part of the application's life cycle.

>In some near future work we're going to be reworking how we start processes so they are no longer a child process of the supervisor. We'll be creating another process - the launcher - which the supervisor communicates with to spawn a child process. This will allow us to upgrade the supervisor without needing to restart PID 1. This also means that the supervisor will now be supervisoring processes a little bit differently, it'll need to read the PID file from the svc directory and then watch the process table/wait for signals from the launcher which it'll proxy to the supervisor.

>By writing to the PID file in the svc directory from the reload hook with the PID of the newly spawned process, this might mean we just get PID swapping for free. What do you think?

This issue can be revisited once the mentioned rework of process handling is complete.

**Comment by christophermaier** [comment]
This _may_ have been fixed with the Launcher, but we'll need to try and reproduce it. Filing this away as an issue to resolve for a 1.0 Supervisor.

**Comment by davidMcneil** [comment]
How would you go about reproducing this issue? This is my attempt. I have a simple package with the following `run` hook

```bash
#!/bin/bash

echo ">>> TEST <<<"
echo "Message is '{{cfg.msg}}'"

while true; do
    echo "Message is still '{{cfg.msg}}'"
    pid=$(<{{pkg.svc_pid_file}})
    echo "Run Hook: pid is '$pid' set in '{{pkg.svc_pid_file}}'"
    sleep 5
done

```
and a `reconfigure` hook
```bash
#!/bin/bash

pid=$(<{{pkg.svc_pid_file}})
echo "Reconfigure Hook: OLD pid='${pid}' set in {{pkg.svc_pid_file}}"

# Start another run hook and set the new PID
{{pkg.svc_path}}/hooks/run &
echo $! >{{pkg.svc_pid_file}}

# Kill the old PID
kill ${pid}

pid=$(<{{pkg.svc_pid_file}})
echo "Reconfigure Hook: NEW pid is '$pid' set in '{{pkg.svc_pid_file}}'"
```
This is the state of the relevant processes before applying a new configuration
```
root      3847  2223  0 11:21 pts/2    00:00:00 sudo env HAB_SUP_BINARY=./target/debug/hab-sup ./target/debug/hab sup run -s at-once
root      3848  3847  0 11:21 pts/2    00:00:00 /hab/pkgs/core/hab-launcher/11300/20190605211433/bin/hab-launch run -s at-once
root      3852  3848  1 11:21 pts/2    00:00:00 ./target/debug/hab-sup run -s at-once
hab       3876  3848  0 11:21 pts/2    00:00:00 /bin/bash /hab/svc/test/run
hab       3991  3876  0 11:22 pts/2    00:00:00 sleep 5
```
These are the logs I get when I apply a new configuration
```
test.default(O): Message is still 'New Message'
test.default(O): Run Hook: pid is '3876' set in '/hab/svc/test/PID'
test.default(O): Message is still 'New Message'
test.default(O): Run Hook: pid is '3876' set in '/hab/svc/test/PID'
hab-sup(CMD): Setting new configuration version 30 for test.default
test.default(CF): Modified configuration file /hab/svc/test/config/config.toml
test.default hook[reconfigure]:(HK): Reconfigure Hook: OLD pid='3876' set in /hab/svc/test/PID
test.default hook[reconfigure]:(HK): Reconfigure Hook: NEW pid is '4062' set in '/hab/svc/test/PID'
test.default hook[reconfigure]:(HK): >>> TEST <<<
test.default hook[reconfigure]:(HK): Message is 'New Message'
test.default hook[reconfigure]:(HK): Message is still 'New Message'
test.default hook[reconfigure]:(HK): Run Hook: pid is '4062' set in '/hab/svc/test/PID'
hab-launch(SV): Child for service 'test.default' with PID 3876 exited with code signal: 15
test.default hook[reconfigure]:(HK): Message is still 'New Message'
test.default hook[reconfigure]:(HK): Run Hook: pid is '4062' set in '/hab/svc/test/PID'
test.default hook[reconfigure]:(HK): Message is still 'New Message'
test.default hook[reconfigure]:(HK): Run Hook: pid is '4062' set in '/hab/svc/test/PID'
```
This is the state of the relevant processes before applying a new configuration
```
root      3847  2223  0 11:21 pts/2    00:00:00 sudo env HAB_SUP_BINARY=./target/debug/hab-sup ./target/debug/hab sup run -s at-once
root      3848  3847  0 11:21 pts/2    00:00:00 /hab/pkgs/core/hab-launcher/11300/20190605211433/bin/hab-launch run -s at-once
root      3852  3848  0 11:21 pts/2    00:00:00 ./target/debug/hab-sup run -s at-once
hab       4060  3852  0 11:22 pts/2    00:00:00 [reconfigure] <defunct>
hab       4062     1  0 11:22 pts/2    00:00:00 /bin/bash /hab/svc/test/hooks/run
hab       4072  4062  0 11:22 pts/2    00:00:00 sleep 5
```

Based on this example it does not seem to work quite right. Although, this may be due to a deficiency in my bash knowledge. A few problems:
* We still get the message about the service dying even though we changed the PID before it died (at least the launcher did not try to restart it)
*

**Comment by davidMcneil** [comment]
This is currently not possible. #6668 resolves this by adding the following documentation:

> Habitat does not support changing the `PID` of the underlying service in any lifecycle hook. If part of a service's reconfiguration relies on changing the `PID`, you should not provide a `reconfigure` hook, and instead, use the default behavior of restarting the service for reconfiguration.
